src/virtualenv/seed/embed/via_app_data/pip_install/base.py

```diff
@@ -31,6 +31,16 @@ def _sync(self, src, dst):
 
     def install(self, version_info):
         self._extracted = True
+        # cleanup existing version's .dist-info if exists
+        # TODO: This still could leave extra packages hanging around but those should be inert
+        dist_name = self._dist_info.stem[: self._dist_info.stem.index("-")]
+        for filename in self._creator.purelib.iterdir():
+            if filename.name.startswith(dist_name) and filename.suffix == ".dist-info" and filename.is_dir():
+                logging.debug(
```

Comment:
I don't think we should log this debug message. From what you're saying it's expected behaviour.

Reply:
> Let's not leave todo-s, we should fix them now. 

Ok, implementing "uninstall" functionality then.

> I don't think we should log this debug message. From what you're saying it's expected behaviour.

Debug message is there to indicate that the conditional expected behavior did in fact occur. This will become especially important since uninstall is a multi-step process.

Reply:
I don't follow, can you detail?

Reply:
Which part? :smile: I've added a commit (without tests) that implements the above logic by removing `<top-level-pkg>/*` as well as everything in `RECORD` file that isn't removed otherwise and then everything that is in `dist-info`. This should take care of everything stale from an existing project.